### PR TITLE
use only relative error in control_plots test

### LINKED ISSUE #26728: Control test fails with numpy 2.0
CC @akshanshbhatt @faze-geek 

This is the second time I have asked to fix these tests (gh-23128).

This test fails with numpy 2.0:
```console
$ pytest sympy/physics/control/tests/test_control_plots.py::test_pole_zero
================================================================================================ test session starts =================================================================================================
platform darwin -- Python 3.12.2, pytest-8.1.1, pluggy-1.4.0
architecture: 64-bit
cache:        yes
ground types: flint (python-flint==0.6.0)

rootdir: /Users/enojb/work/dev/sympy
configfile: pyproject.toml
plugins: instafail-0.5.0, doctestplus-1.2.2.dev5+g5176f32, hypothesis-6.99.9, xdist-3.5.0, timeout-2.3.1, split-0.8.2
collected 1 item

sympy/physics/control/tests/test_control_plots.py F                                                                                                                                                            [100%]

====================================================================================================== FAILURES ======================================================================================================
___________________________________________________________________________________________________ test_pole_zero ___________________________________________________________________________________________________

    def test_pole_zero():
        if not numpy:
            skip("NumPy is required for this test")

        def pz_tester(sys, expected_value):
            z, p = pole_zero_numerical_data(sys)
            z_check = numpy.allclose(z, expected_value[0])
            p_check = numpy.allclose(p, expected_value[1])
            return p_check and z_check

        exp1 = [[], [-0.24999999999999994+1.3919410907075054j, -0.24999999999999994-1.3919410907075054j]]
        exp2 = [[0.0], [-0.25+0.3227486121839514j, -0.25-0.3227486121839514j]]
        exp3 = [[0.0], [-0.5000000000000004+0.8660254037844395j,
            -0.5000000000000004-0.8660254037844395j, 0.9999999999999998+0j]]
        exp4 = [[], [5.0, 0.0, 0.0, 0.0]]
        exp5 = [[-5.645751311064592, -0.5000000000000008, -0.3542486889354093],
            [-0.24999999999999986+1.3919410907075052j,
            -0.24999999999999986-1.3919410907075052j, -0.2499999999999998+0.32274861218395134j,
            -0.2499999999999998-0.32274861218395134j]]
        exp6 = [[], [-1.1641600331447917-3.545808351896439j,
              -0.8358399668552097+2.5458083518964383j]]

>       assert pz_tester(tf1, exp1)
E       assert False
E        +  where False = <function test_pole_zero.<locals>.pz_tester at 0x1148e6c00>(TransferFunction(1, p**2 + 0.5*p + 2, p), [[], [(-0.24999999999999994+1.3919410907075054j), (-0.24999999999999994-1.3919410907075054j)]])

sympy/physics/control/tests/test_control_plots.py:121: AssertionError
                                                                                                   DO *NOT* COMMIT!
============================================================================================== short test summary info ===============================================================================================
FAILED sympy/physics/control/tests/test_control_plots.py::test_pole_zero - assert False
================================================================================================= 1 failed in 0.29s ==================================================================================================
```

### PR DESCRIPTION
**Pull request by smichr**
fixes #26728

#### Release Notes

NO ENTRY

### CODE REVIEW

**Comment by oscarbenjamin** [comment]
We need to make sure that this test is actually run in CI.

**Comment by smichr** [comment]
> We need to make sure that this test is actually run in CI.

Can you clarify? It failed (per issue) so doesn't that mean it is being run?

**Comment by oscarbenjamin** [comment]
> It failed (per issue) so doesn't that mean it is being run?

No, it failed because I ran the tests locally. The test will be skipped if numpy is not installed which is the default mode for running the tests in CI.

The optional dependency jobs test with numpy but don't run the whole test suite.

**Comment by smichr** [comment]
> if numpy is not installed which is the default mode

So are you saying that we need to enable numpy as a dependency?  (I tested the failing routine locally and it passed.)

**Comment by oscarbenjamin** [comment]
> So are you saying that we need to enable numpy as a dependency?

We need to make sure that this test runs in the CI jobs that have numpy installed. I'm not sure why it doesn't run:
https://github.com/sympy/sympy/blob/6d6e89caad6da38313a9174fc186b67904872a47/bin/test_optional_dependencies.py#L27

**Comment by akshanshbhatt** [comment]
> No, it failed because I ran the tests locally. The test will be skipped if numpy is not installed which is the default mode for running the tests in CI.

From what I can see in the CI logs, this test is skipped specifically for the numpy nightly/prerelease versions. It passes in all the other `optional-dependencies` environments.


Screenshot 2024-06-23 at 1 05 09 PM

**Comment by smichr** [comment]
I don't know what needs to be done to make sure it is run in CI. I see other lines of the file specifically reference the tests directory not only the parent directory. I haven't looked into how this works wrt full or partial paths.

**Comment by oscarbenjamin** [comment]
> I don't know what needs to be done to make sure it is run in CI.

I think it actually is running in CI. The test failure was seen locally on MacOS and is due to the roots being returned in a different order by `np.roots` when a different BLAS library is used: https://github.com/numpy/numpy/issues/26772#issuecomment-2184972180.

**Comment by oscarbenjamin** [comment]
The changed test here still fails for me on MacOS (ARM).

The problem is that the roots returned by np.roots are in a different order: https://github.com/sympy/sympy/issues/26728#issuecomment-2184950317

**Comment by smichr** [comment]
> are in a different order

Could you test again? They are now canonical, but failing tests and doctests will need to be updated.

**Comment by oscarbenjamin** [comment]
> Could you test again?

It still fails:
```
> /Users/enojb/work/dev/sympy/sympy/physics/control/tests/test_control_plots.py(129)test_pole_zero()
-> assert pz_tester(tf1, exp1)
(Pdb) p pole_zero_numerical_data(tf1)
(array([], dtype=float64), array([-0.25-1.39194109j, -0.25+1.39194109j]))
```

File: sympy/physics/control/control_plots.py [hunk 1]
```diff
@@ -88,8 +90,8 @@ def pole_zero_numerical_data(system):
     >>> from sympy.physics.control.lti import TransferFunction
     >>> from sympy.physics.control.control_plots import pole_zero_numerical_data
     >>> tf1 = TransferFunction(s**2 + 1, s**4 + 4*s**3 + 6*s**2 + 5*s + 2, s)
-    >>> pole_zero_numerical_data(tf1)   # doctest: +SKIP
-    ([-0.+1.j  0.-1.j], [-2. +0.j        -0.5+0.8660254j -0.5-0.8660254j -1. +0.j       ])
+    >>> pole_zero_numerical_data(tf1)
```
Comment (smichr) [inline sympy/physics/control/control_plots.py:93]:
```suggestion
    >>> pole_zero_numerical_data(tf1)  # doctest: +SKIP
```

File: sympy/physics/control/control_plots.py [hunk 2]
```diff
@@ -109,7 +111,12 @@ def pole_zero_numerical_data(system):
     zeros = np.roots(num_poly)
     poles = np.roots(den_poly)
 
-    return zeros, poles
+    # make ordering canonical
+    def _sort(l):
+        return [float(i) if i.is_real else complex(i) for i in
+                _nsort([_sympify(i) for i in l])]
```
Comment (smichr) [inline sympy/physics/control/control_plots.py:117]:
```suggestion
                _nsort([_sympify(i) for i in l])] if len(l) else []
```

**Comment by smichr** [comment]
I made `_nsort` handle the empty list and tested mods locally. Could you check once more, please?

**Comment by oscarbenjamin** [comment]
Now it passes for tf1 but fails for tf3:
```
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> PDB post_mortem (IO-capturing turned off) >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
> /Users/enojb/work/dev/sympy/sympy/physics/control/tests/test_control_plots.py(131)test_pole_zero()
-> assert pz_tester(tf3, exp3)
(Pdb) p pole_zero_numerical_data(tf3)
([0.0], [(-0.5000000000000001-0.866025403784439j), (-0.5000000000000001+0.866025403784439j), (1-1.3256180803817732e-34j)])
```

File: sympy/physics/control/tests/test_control_plots.py [hunk 3]
```diff
@@ -101,20 +102,27 @@ def test_pole_zero():
...
+        def check(a, b):
+            if isinstance(a, (list, tuple)):
+                return all(check(i, j) for i,j in zip(a, b))
+            a = sympify(a).n(chop=1e-10)
+            if not b:
+                return all_close(a, b, atol=1e-10, rtol=0)
+            return all_close(a, b, atol=0, rtol=1e-6)
+        z_check = check(z, expected_value[0])
+        p_check = check(p, expected_value[1])
         return p_check and z_check
```
Comment (smichr) [inline sympy/physics/control/tests/test_control_plots.py:115]:
```suggestion
        zp = pole_zero_numerical_data(sys)
        zp = [[_sympify(i).n(chop=True) for i in j] for j in zp]
        z, p = [_sort(i) for i in zp]
        def check(a, b):
            if isinstance(a, (list, tuple)):
                return all(check(i, j) for i,j in zip(a, b))
            if not b:
                return all_close(a, b, atol=1e-10, rtol=0)
            return all_close(a, b, atol=0, rtol=1e-6)
        z_check = check(z, expected_value[0])
        p_check = check(p, expected_value[1])
        return p_check and z_check
```

File: sympy/physics/control/tests/test_control_plots.py [hunk 4]
```diff
@@ -1,8 +1,10 @@
 from math import isclose
-from sympy.core.numbers import I
+from sympy.core.numbers import I, all_close
+from sympy.core.sympify import sympify
```
Comment (smichr) [inline sympy/physics/control/tests/test_control_plots.py:3]:
```suggestion
from sympy.core.sympify import _sympify
```

**Comment by smichr** [comment]
It will now chop before sorting. For purposes of this test, instead of chopping we could sort by magnitude and then by real and imaginary parts, but I suspect it will pass now.

File: sympy/physics/control/tests/test_control_plots.py [hunk 5]
```diff
@@ -101,20 +103,28 @@ def test_pole_zero():
         skip("NumPy is required for this test")
 
     def pz_tester(sys, expected_value):
-        z, p = pole_zero_numerical_data(sys)
-        z_check = numpy.allclose(z, expected_value[0])
-        p_check = numpy.allclose(p, expected_value[1])
+        zp = pole_zero_numerical_data(sys)
+        zp = [[_sympify(i).n(chop=True) for i in j] for j in zp]
```
Comment (smichr) [inline sympy/physics/control/tests/test_control_plots.py:107]:
```suggestion
        zp = [[_sympify(i).n(chop=1e-10) for i in j] for j in zp]
```

**Comment by oscarbenjamin** [comment]
This now passes locally on MacOS. Thanks.

**Pull request merged**